### ISSUE TITLE
Incorrect HttpClient.Timeout

### ISSUE THREAD

**Issue by spouliot** [mono/mono#12100]
While trying to fix https://github.com/xamarin/xamarin-macios/issues/5190 I was surprised by mono (5.18) behaviour. I tested it on .netcore and it behave like I thought it should.

## Steps to Reproduce

1. Create `x.cs` with the following code

```csharp
using System;
using System.Net.Http;
using System.Threading.Tasks;

namespace httpclient
{
	class Program
	{
		static async Task Test (HttpMessageHandler handler)
		{
			Console.WriteLine ($"### {handler}");
			var now = DateTime.UtcNow;
			try {
				var client = new HttpClient (handler);
				client.Timeout = TimeSpan.FromSeconds (150);
				HttpResponseMessage response = await client.GetAsync ("https://httpstat.us/200?sleep=120000");
				response.EnsureSuccessStatusCode ();
				Console.WriteLine ("PASS");
			} catch (Exception e) {
				Console.WriteLine ($"FAIL {e}");
			} finally {
				Console.WriteLine ($"Time {(DateTime.UtcNow - now).TotalSeconds}");
			}
		}

		static void Main (string [] args)
		{
			var handler = new HttpClientHandler ();
			Test (handler).Wait ();
		}
	}
}
```

2. Compile provided source code

```
$ csc x.cs -r:System.Net.Http.dll
Microsoft (R) Visual C# Compiler version 2.8.2.62916 (2ad4aabc)
Copyright (C) Microsoft Corporation. All rights reserved.
```

3. Execute it with mono 5.18 - wait for more than 150 seconds (it must be more than 100 since mono seems hardcoded to that value).

### Current Behavior

```
$ mono x.exe
### System.Net.Http.HttpClientHandler
FAIL System.Net.Http.HttpRequestException: An error occurred while sending the request ---> System.Net.WebException: The operation has timed out.
  at System.Net.HttpWebRequest.RunWithTimeoutWorker[T] (System.Threading.Tasks.Task`1[TResult] workerTask, System.Int32 timeout, System.Action abort, System.Func`1[TResult] aborted, System.Threading.CancellationTokenSource cts) [0x000f8] in <d012d9eca6d14975a47488863de2f4c6>:0 
  at System.Net.HttpWebRequest.EndGetResponse (System.IAsyncResult asyncResult) [0x00020] in <d012d9eca6d14975a47488863de2f4c6>:0 
  at System.Threading.Tasks.TaskFactory`1[TResult].FromAsyncCoreLogic (System.IAsyncResult iar, System.Func`2[T,TResult] endFunction, System.Action`1[T] endAction, System.Threading.Tasks.Task`1[TResult] promise, System.Boolean requiresSynchronization) [0x0000f] in <96207d0baa204f48a53ad6be05f5ecba>:0 
--- End of stack trace from previous location where exception was thrown ---

  at System.Net.Http.HttpClientHandler.SendAsync (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x0041d] in <484e1189bad34d1ca999fcd9f244eb40>:0 
   --- End of inner exception stack trace ---
  at System.Net.Http.HttpClientHandler.SendAsync (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x00478] in <484e1189bad34d1ca999fcd9f244eb40>:0 
  at System.Net.Http.HttpClient.SendAsyncWorker (System.Net.Http.HttpRequestMessage request, System.Net.Http.HttpCompletionOption completionOption, System.Threading.CancellationToken cancellationToken) [0x000ca] in <484e1189bad34d1ca999fcd9f244eb40>:0 
  at httpclient.Program.Test (System.Net.Http.HttpMessageHandler handler) [0x000d1] in <6472dc270f0b42e58d2c9a68c7d0d169>:0 
Time 100.149279
```

Note that the timeout occurs after 100 seconds, not the 150 seconds that was set on the `HttpClient`.

### Expected Behavior

No timeout - since we asked for a 120 seconds delay (which is under the 150 seconds timeout).

```
### System.Net.Http.HttpClientHandler
PASS
Time 120.907851
```

## On which platforms did you notice this

[x] macOS
[ ] Linux
[ ] Windows

**Version Used**:

```
Mono JIT compiler version 5.18.0.185 (2018-08/ea740d2bc8d Tue Oct 30 08:15:35 EDT 2018)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          yes(600)
	Suspend:       preemptive
	GC:            sgen (concurrent by default)
```

**Comment by ValdasD** [comment]
Hi, is there any updates or workarounds on this? Experiencing the same issue on Linux.

**Comment by marek-safar** [comment]
@baulig any word on this one?

**Comment by baulig** [comment]
It is behaving as intended, setting the `HttpClient.Timeout` property only controls how long its async methods should wait for completion.  It has no effect on the underlying connection.

Closing as this is not a Mono bug.  If you don't like the behavior, then please discuss it with CoreFx as their implementation is doing the exact same thing.

**Comment by tomaxsas** [comment]
it's valid bug, mono 5.4.1.7 works fine.  https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient.timeout?view=netframework-4.7.2 here is docs for property. same code on .net core on linux works fine, as well as mono 5.4.1.7, but not >= 5.12 version of mono

**Comment by baulig** [comment]
I'm going to test this with our new code.

**Comment by baulig** [comment]
This is actually working just fine when using the new `HttpClient` from CoreFX.

**Comment by marek-safar** [comment]
@baulig could you check what was the change which caused this?

**Comment by tomaxsas** [comment]
any ETA on this? tried 5.20.0.237, still no luck. 5.1.4.7 works fine.

**Comment by tomaxsas** [comment]
Tested with  6.1.0 nightly, problem solved in this  at least on Linux (not tired on other paltforms)

**Comment by DevEddy** [comment]
Is this included in `Mono 6.4.0.208 (2019-06/07c23f2ca43) (64-bit)`? 
I am experiencing this only on Android (iOS and UWP is OK). 

```
Xamarin.Android
Version: 10.0.3.0 (Visual Studio Community)
Commit: xamarin-android/d16-3/4d45b41
```
